File: apps/api/models/customer.py
"""
Customer model – represents individual or business clients.

Schema reference: docs/ARCHITECT.md § Model: Customer
Dictionary:       docs/DICTIONARY.md  (balance = công nợ)

Balance semantics
-----------------
  balance > 0  →  Customer owes money  (debt / công nợ dương)
  balance < 0  →  Customer has credit  (over-paid / công nợ âm)
  balance == 0 →  Settled
"""

import uuid
from typing import List, Optional

from sqlmodel import Field, Relationship, SQLModel

from .enums import CustomerType
from .ticket import Ticket
from .transaction import Transaction

# ---------------------------------------------------------------------------
# Shared / base properties
# ---------------------------------------------------------------------------

class CustomerBase(SQLModel):
    """Fields shared between create/read schemas and the DB table."""

    name: str = Field(
        index=True,
        min_length=1,
        max_length=255,
        description="Full name of the individual or registered business name.",
    )
    type: CustomerType = Field(
        default=CustomerType.INDIVIDUAL,
        description="INDIVIDUAL for personal travellers; BUSINESS for corporate accounts.",
    )
    balance: float = Field(
        default=0.0,
        description=(
            "Current debt balance (công nợ). "
            "Positive = customer owes; Negative = customer has credit."
        ),
    )


# ---------------------------------------------------------------------------
# DB Table
# ---------------------------------------------------------------------------

class Customer(CustomerBase, table=True):
    """Persisted customer entity stored in the `customer` table."""

    __tablename__ = "customer"

    id: Optional[uuid.UUID] = Field(
        default_factory=uuid.uuid4,
        primary_key=True,
        index=True,
        nullable=False,
    )

    # Relationships – populated lazily by SQLAlchemy; not included in API responses by default.
    tickets: List[Ticket] = Relationship(back_populates="customer")
    transactions: List[Transaction] = Relationship(back_populates="customer")


# ---------------------------------------------------------------------------
# Request / Response schemas
# ---------------------------------------------------------------------------

class CustomerCreate(CustomerBase):
    """Payload accepted by POST /customers."""
    pass


class CustomerRead(CustomerBase):
    """Public representation of a customer."""

    id: uuid.UUID


class CustomerDirectoryItem(SQLModel):
    """Slim customer payload used by the directory page."""

    id: uuid.UUID
    full_name: str
    phone: Optional[str] = None
    current_balance: float


class CustomerUpdate(SQLModel):
    """All fields optional for partial PATCH payloads."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    type: Optional[CustomerType] = None
    balance: Optional[float] = None

File: apps/web/src/schemas/customer.ts
/**
 * customer.ts – Zod schemas for the Customer domain.
 *
 * Mirrors apps/api/models/customer.py (CustomerCreate, CustomerRead, CustomerUpdate).
 *
 * Balance semantics (công nợ) – from docs/DICTIONARY.md:
 *   balance > 0  →  Customer owes money  (debt)
 *   balance < 0  →  Customer has credit  (over-paid)
 *   balance === 0 → Settled
 */

import { z } from "zod";
import { CustomerTypeSchema } from "./enums";

// ---------------------------------------------------------------------------
// Shared base
// ---------------------------------------------------------------------------

const CustomerBaseSchema = z.object({
  /** Full name of the individual or registered business. Maps to Python: name */
  name: z
    .string()
    .min(1, "Name is required.")
    .max(255, "Name must be at most 255 characters."),

  /** Account category. Maps to Python: type */
  type: CustomerTypeSchema.default("INDIVIDUAL"),

  /**
   * Current debt balance (công nợ).
   * Positive = customer owes; Negative = customer has credit.
   * Maps to Python: balance
   */
  balance: z.number().default(0),
});

// ---------------------------------------------------------------------------
// CustomerCreate – POST /customers
// ---------------------------------------------------------------------------

export const CustomerCreateSchema = CustomerBaseSchema;
export type CustomerCreate = z.infer<typeof CustomerCreateSchema>;

// ---------------------------------------------------------------------------
// CustomerRead – GET /customers/:id
// ---------------------------------------------------------------------------

export const CustomerReadSchema = CustomerBaseSchema.extend({
  /** UUID assigned by the backend. Read-only on the frontend. */
  id: z.string().uuid(),
});

export type CustomerRead = z.infer<typeof CustomerReadSchema>;

// ---------------------------------------------------------------------------
// CustomerUpdate – PATCH /customers/:id (all fields optional)
// ---------------------------------------------------------------------------

export const CustomerUpdateSchema = z.object({
  name: z
    .string()
    .min(1, "Name is required.")
    .max(255, "Name must be at most 255 characters.")
    .optional(),

  type: CustomerTypeSchema.optional(),

  /** Direct balance override – normally mutated via Transactions only. */
  balance: z.number().optional(),
});

export type CustomerUpdate = z.infer<typeof CustomerUpdateSchema>;

export const LedgerEntrySchema = z.object({
  id: z.string().uuid(),
  entry_type: z.enum(["ticket", "payment", "adjustment"]),
  created_at: z.coerce.date(),
  content: z.string(),
  amount: z.number(),
  running_balance: z.number(),
});

export type LedgerEntry = z.infer<typeof LedgerEntrySchema>;

export const CustomerLedgerSchema = z.object({
  customer: CustomerReadSchema,
  current_balance: z.number(),
  entries: z.array(LedgerEntrySchema),
});

export type CustomerLedger = z.infer<typeof CustomerLedgerSchema>;

export const RecordPaymentSchema = z.object({
  amount: z.number().positive("Amount must be a positive number."),
  note: z.string().max(500, "Note must be at most 500 characters.").optional(),
});

export type RecordPayment = z.infer<typeof RecordPaymentSchema>;

export const CustomerDirectoryItemSchema = z.object({
  id: z.string().uuid(),
  full_name: z.string(),
  phone: z.string().nullable().optional(),
  current_balance: z.number(),
});

export type CustomerDirectoryItem = z.infer<typeof CustomerDirectoryItemSchema>;
